</br><font color = #000066 size=6.5> Facultad de Ciencias</font> 

<br><font color = #CC9933 size=6.5> UNAM </font> 

 <br><font color = silver size=7> Programación en Python</font> 

***

# Notas del curso Métodos Cuantitativos en Finanzas (Parte 1)


El presente curso de programación está dividido en partes para mantener más legible, entendible y ordenado el contenido. 

<br><font color = #708090 size=1> El contenido de este documento fue inspirado en las notas de los cursos *Python for Finance* e *Introduction to Python Primer* impartidos por los doctores Antoine Jacquier, Aitor Murguraza Gonzalez y Riaz Ahmad. Para cualquier duda o comentario el contenido, escribir a marianne.toscano@ciencias.unam.mx  </font>  

## Introducción a Python

### ¿Por qué Python?

* De acuerdo con *eFinancialCareers*, Python es el lenguaje de programación que te hará conseguir un trabajo en el sistema bancario - *The coding languages that will get you a job in banking, ranked* https://www.efinancialcareers.co.uk/news/finance/the-six-hottest-programming-languages-to-know-in-banking-technology

<img src="./img/current_job_openings.png" style="width:500px">

* *Python was also beginning to garner attention from the scientific community, particularly in neuroscience research, where it started to be used extensively. With the development of the library NumPy, Python started to be seen as an alternative to R, Matlab, and other scientific computing languages for research projects.* https://www.welcometothejungle.com/en/articles/btc-python-popular


* *Python: El lenguaje del futuro* https://www.baoss.es/python-el-lenguaje-del-futuro/


* La UNAM también difunde y promueve el aprendizaje de python https://unamglobal.unam.mx/la-unviersidad-de-michigan-ofrece-popular-curso-de-python-gratis-y-en-espanol/


La historia de Python en wikipedia.   https://en.wikipedia.org/wiki/History_of_Python. La imagen siguiente ilustra una serie de televisión en la que se inspiraron los creadores de este leguaje para ponerle nombre.


<img src="./img/Monty_pythons_flying_circus.jpeg" style="width:500px">


Programamos por **necesidad**. La complexidad de muchos problemas de las áreas laborales, académicas o de investigación nos impide que utilicemos solo técnicas analyticas para modelar. Nos vemos muy limitados. Por ejemplo, en finanzas, a menos de que el problema sea muy básico om que incluya muchos supuestos, es altamente improbable que exista una solución de forma cerrada.

###  Características del lenguaje de programación
* Es simple para entender y fácil de recordar 
* El código se puede leer como si se estuviera leyendo algo escrito en inglés
* Es código abierto 

>* Libre distribución
>* El código fuente de todas las funciones debe estar incluido
>* Cualquiera puede modificar/mejorar el código fuente
>* Puede ejecutarse en diferentes dispositivos y sistemas operativos (Windows, IOS, Linux)

* Popularmente usardo para Machine Learning y Data Science
* Lenguaje interpretado

>* A diferencia de los lenguajes compilados, son más lentos
>* Más facil para aprender
>* Menos estricto para los errores de programación

* Alta polularidad y gran apoyo de la comunidad
* Tiene una selección extensiva de bibliotecas y paquetes 
>* Keras, TensorFlow, Scikit-learn para machine learning
>* NumPy para para computación científica de alto rendimiento y análisis de datos 
>* SciPy para computacióm científica avanzada
>* Pandas para análisis de datos 
>* Seaborn para (fancy) visualización de datos



***

## Primeros pasos

### Distribuciones de Python

Python es un lenguaje de programación, entonces existen diferentes plataformas de distribución, compiladores e intérpretes para dar forma a los códigos y desarrollar los programas. Como ejemplos tenemos: Python official, Anaconda Python, Enthought Canopy, pythonxy,...

Cada distribuidor proporciona diferentes y útiles herramientas, como editores, paqueterías preinstaladas, soporte en algunas plataformas.



En Linux y Mac OS, Python viene pre-instalado con el sistema operativo, no obstante, muchas paquetes o módulos (por ejemplo, SciPy) se deben instalar manualmente. 

<img src="./img/linux_macOS_windows.png" style="width:200px">

### Paso 1. Iniciamos con Anaconda
<img src="./img/Anaconda_logo.png" style="width:200px">

Usaremos Anaconda por su fácil y amigable interfaz https://www.anaconda.com/download/



Por facilidad, Anaconda es la herramienta que usaremos. Anaconda es una plataforma de código abierto que tiene diversas herramientas; incluyendo una distribución de Python, un administrador de paquetes conda, una forma de administrar el entorno y muchos módulos y paquetes preinstalados (por ejemplo, cuadernos NumPy, SciPy y Jupyter). 

<img src="./img/Anaconda_platform.png" style="width:300px">

El uso de Anaconda facilita enormemente el proceso de aprendizaje y evita muchos problemas a los que se enfrentan los nuevos programadores de Python. Más importante aún, es excelente para enseñar en el aula.

<img src="./img/Anaconda.png" style="width:500px">

#### Paso 1.1 Instalación de Anaconda

Para instalar Anaconda, primero tenemos que saber si la computadora o máquina con la que voy a trabajr tiene un procesador de 32 bits o de 64 bits.

##### ¿Cómo saber si mi PC/laptop es de 32 o 64-bit?

* **Linux** uname -a  (si ves x64 entonces 64-bit, si no, 32-bit)
* **Mac OS** uname -a
* Windows (viejito) - Click en Start > click derecho en My Computer > click en Properties > dirá 64-bit 
* **Windows 10** - En Search escribe "device specifications" > System Specifications


Ahora instala de acuerdo a tu sistema operativo

<img src="./img/Anaconda_installers.png" style="width:600px">

Usaremos la versión *Python 3*

### Paso 2.  Inicialización de Jupyter Notebook

Una vez instalado todo, puedes ingresar a `Jupyter notebook` por diversas maneras, las más sencillas son las siguientes:

* **Mac OS** En la terminal escribe `jupyther notebook`
<img src="./img/Mac_terminal_jupyter.png" style="width:600px">


* **Linux** En la terminal escribe `jupyther notebook`
<img src="./img/Linux_terminal_jupyter.png" style="width:600px">


* **Windows** En Anaconda prompt escribe `jupyther notebook`
<img src="./img/Windows_terminal_jupyter.png" style="width:600px">


* **Desde el Navegador de Anaconda** da click en el botón *Lounch Jupyter*
<img src="./img/Anaconda_navigator.png" style="width:600px">


Jupyter se ejecuta en su navegador web, pero no requiere una conexión a Internet para funcionar. Python y el IDE se ejecutan localmente en tu máquina, el navegador es solo una interfaz de usuario conveniente.

<img src="./img/jupyter_inicio.png" style="width:600px">

### Paso 3. Crear un notebook de Jupyter de Python

Luego de instalar Anaconda y abrir Jupyter, un nuevo notebook puede abrirse seleccionando *New* > *Python 3*

<img src="./img/new_python3.png" style="width:400px">

<br><font color = #BB9963 size=5> ¡ojo! </font>  Dependiendo de la versión de Anaconda que instales (o ya tengas), habrán más opciones como `Python 2`o `Python 3`. La elección que hagas aquí define el lenguaje de programación del notebook. **En este curso, siempre usaremos Python 3**.

Se pueden instalar otros lenguajes de programación en Jupyter. 

¿Cómo? con el comando `conda install -c r r-irkernel` en tu terminal o en Anaconda prompt. Una vez instalado, verás `R`en el menu *New*.
<img src="./img/R_logo.png" style="width:100px">


***

## Introducción de IPython notebook

### 1. Celdas `Cell` 
Una celda es como un ladrillo con el que se contruyen los notebooks de jupyter. Cada notebook consiste en múltiples celdas y se pueden añadir con hacer click en `Insert` o click en el botón con el símbolo `+`
<img src="./img/insert_cell.png" style="width:400px">

Para eliminar celdas le puedes dar click en `Edit`y luego seleccionar `Delete Cells`

#### 1.1 Tipos de celdas

En jupyther notebook hay fundamentalmente dos tipos de celdas:
* `Markdown` se usa para escribir texto y formulas matemáticas como  


$$
\frac{1}{2\pi}\int_{\mathbb{R}}\exp\left(-\frac{x^2}{2}\right)dx = \cdots
$$

<br><font color = red size=5> ¡ojo! </font>  Se recomienda que, siempre que entregues tareas en notebooks de python (y para aplicar a trabajos que soliciten códigos, para hacer tareas en equipo, proyectos, etc.), utilices `Markdown`. Esto hará más comprensible tu código.


* `Code` se utiliza para ejecutar código

In [1]:
print("¡Hola estudiantes de Ciencias!", "Este es un ejemplo de celda tipo Code.")

¡Hola estudiantes de Ciencias! Este es un ejemplo de celda tipo Code.


#### 1.2 Modos de ejecución de las celdas

Los notebooks tienen dos modos de ejecución de las celdas

1. El modo de comando - la celda se pinta de un marco azul (tecla `Esc`). En ese modo se pueden añadir, borrar celdas, utilizar shortcuts.

2. El modo de edición - la celdas se pintan de un marco verde (tecla `Enter`).

Presionando con doble click en una celda puedes ver su contenido, ya sea código o contenido en Markdown.

### 2. El Kernel

Una sesión o un *kernel* gestiona todo el código, las variables, los nombres,..., dentro de un notebook. El *kernel* puede ser reestablecido en cualquier momento de ser necesario 

<img src="./img/kernel_menu.png" style="width:400px">

### 3. Python, versión y documentación

Como practicantes de matemáticas, no hay necesidad de moemorizar decenas de resultados de cálculo diferencial e integral para ocupar esas técnicas, con tal de que sepamos los resultados básicos y los entendamos, para todo lo demás hay tablas o el propio *Google*. 

Lo mismo pasa con programar, nadie tiene el tiempo de memorizar cada uno de los funciones de Python y muy problablemente hay mucho que nunca tendrán que utilizar. Así, aparte de una pequeña cantidad de funciones que tendrás que memorizar, el resto las puedes buscar en las siguientes fuentes:

* La principal documentación de Python está disponible en https://docs.python.org/3.8/reference/ 
* Stackoverflow es bastante útil https://stackoverflow.com/
* Googléalo (obviamente)

#### Versión de Python

Para conocer la versión que estamos usando actualmente se ejecuta el código `!python3 --version`. 

In [2]:
!python --version

Python 3.8.3


#### Usar el elemento `!` 

En la celda anterior usamos `!` para conocer la versión que tenemos instalada de Python. `!` le dice al interprete que mande el código a la línea de comando o terminal o Anaconda prompt de mi computador. De esta manera podemos utilizar este tipo de código que se debe correr en la terminal, directamente aquí en una celda tipo `Code`.

### 4. Instalar paquetes/librerías/módulos de Python

Uno de los grandes beneficios de Python es que su comunidad desarrolladora es muy activa. Esto significa que, antes de que pienses en implementar un algoritmo por tu cuenta, vale la pena que revises si exite algun paquete de libre distribucón que realice la actividad que necesitas.

**PIP (Pip Installs Packages)** es la herramienta para instalar dichos paquetes. 

Por ejemplo, YFinance es popular para obtener datos financieros desde Yahoo finance, podemos instalarla con `pip install yfinance` en la línea de comandos, terminal o Anaconda prompt.

O bien, usando la *keyword* `!` en una celda `Code` de la siguiente manera:

In [3]:
!pip install yfinance

#### Short cuts 

Jupyter tiene los siguientes Short cuts, para ahorrar tiempo al programar:

* <mark>Shift + Enter</mark>: Run current cell and take cursor to the next one

* <mark>Ctrl + Enter</mark>: Run current cell

* <mark>Alt + Enter</mark>: Run current cell and insert new cell below

* <mark>Esc+M</mark>: Convert current cell to markdown

* <mark>Esc+Y</mark>: Convert current cell to code

* <mark>Esc + A</mark>: Insert new cell above current cell

* <mark>Esc + B</mark>: Insert new cell below current cell

* <mark>Esc + H</mark>: Show shortcuts menu

In [4]:
432+5

437

In [5]:
94793+543

95336

***

## Operaciones básicas de Python y Tipos de datos

In [6]:
a=5
b=7
c=2.

print(type(a))
print(type(c))

<class 'int'>
<class 'float'>


El operador *potencia* en python se escribe con doble asterisco `** `

In [7]:
3+6

9

In [8]:
print(a+b)
print(a**b) 
print(type(a+b))
print(type(a*c))

12
78125
<class 'int'>
<class 'float'>


Para comentar tu código usa `#`, todo lo que le procesa al sìmbolo *gato* no será ejecutado.

In [9]:
print(a**b) # doble asterisco representa el operador potencia

78125


<br><font color = #BB9963 size=5> ¡ojo! </font> El operador `^` no indica una potencia, pero sí es el operador lógico XOR (veer https://es.wikipedia.org/wiki/Disyunción_exclusiva).

### Keywords

Las *keywords* son caracteres especiales que no pueden ser usados para darle nombre a una variable, ya que ellas tienen una implementación interna

In [10]:
import keyword

print(keyword.kwlist)

['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


### Guía de estilo para programar en Python

PEP 8 -- Style Guide for Python Code (https://www.python.org/dev/peps/pep-0008/) hay convenciones recomendadas para escribir un buen código

*Ejemplo*: Segun las reglas de estilo

esto es correcto:

`foo = long_function_name(var_one, var_two,
                         var_three, var_four)`                         

y esto no es correcto:                        
                        
`foo = long_function_name(var_one, var_two,
    var_three, var_four)`
                         
Desde las notas del curso 'Introduction to Python' del Dr. Antoine Jacquier, esta tabla resume las voncenciones para nombrar variables.

|   Type   |                                                Naming Convention                                                |                 Examples                |
|:--------:|:---------------------------------------------------------------------------------------------------------------:|:---------------------------------------:|
| Function | Use a lowercase word or words. Separate words by underscores to improve readability.                            | function, my_function                   |
| Variable | Use a lowercase single letter, word, or words. Separate words with underscores to improve readability.          | x, var, my_variable                     |
| Class    | Start each word with a capital letter. Do not separate words with underscores. This style is called camel case. | Model, MyClass                          |
| Method   | Use a lowercase word or words. Separate words with underscores to improve readability.                          | class_method, method                    |
| Constant | Use an uppercase single letter, word, or words. Separate words with underscores to improve readability.         | CONSTANT, MY_CONSTANT, MY_LONG_CONSTANT |
| Module   | Use a short, lowercase word or words. Separate words with underscores to improve readability.                   | module.py, my_module.py                 |
| Package  | Use a short, lowercase word or words. Do not separate words with underscores.                                   | package, mypackage                      |

### Importar módulos/paquetes/blibliotecas

Para saber qué módulos hay actualmente en el directorio se ejecuta

In [11]:
print(dir())

['In', 'Out', '_', '_4', '_5', '_7', '__', '___', '__builtin__', '__builtins__', '__doc__', '__loader__', '__name__', '__package__', '__spec__', '_dh', '_exit_code', '_i', '_i1', '_i10', '_i11', '_i2', '_i3', '_i4', '_i5', '_i6', '_i7', '_i8', '_i9', '_ih', '_ii', '_iii', '_oh', 'a', 'b', 'c', 'exit', 'get_ipython', 'keyword', 'quit']


Una característica importante de Python como ecosistema, en comparación con otro lenguaje de programación, es la disponibilidad de una gran cantidad de bibliotecas y herramientas. 

Estas bibliotecas y herramientas generalmente deben **importarse** cuando de necesiten (por ejemplo, una biblioteca para graficar). *Importar* `import` significa hacer que una biblioteca esté disponible para el espacio de nombres actual y el proceso de interpretación de Python actual. 

Python ya viene con un gran conjunto de bibliotecas que mejoran el intérprete básico, por ejemplo, para cálculos matemáticos básicos pyth. Sin embargo, si se requiere de operaciones matemáticas más complejas, deben importarse a través de la biblioteca `math`. 

Cuando se inicia un programa de Python, solo tiene acceso a funciones y clases básicas.
("int", "dict", "len", "sum", "range", ...) 

In [12]:
import matplotlib # graphics library
import scipy # scientific python 
import nltk  # for natural language processing 
import numpy # numerical library and manipulating arrays - flagship
import pandas # panel data - series and dataframes - also flagship

Después de usar cualquier módulo, si ya no lo vamos a ocupar más se puede borrar con la palabra `del`

In [13]:
del nltk
print(dir())

['In', 'Out', '_', '_4', '_5', '_7', '__', '___', '__builtin__', '__builtins__', '__doc__', '__loader__', '__name__', '__package__', '__spec__', '_dh', '_exit_code', '_i', '_i1', '_i10', '_i11', '_i12', '_i13', '_i2', '_i3', '_i4', '_i5', '_i6', '_i7', '_i8', '_i9', '_ih', '_ii', '_iii', '_oh', 'a', 'b', 'c', 'exit', 'get_ipython', 'keyword', 'matplotlib', 'numpy', 'pandas', 'quit', 'scipy']


Se recomienda altamente usar la notación de objetos. Esa práctica le permite al programador tener mayor control y hacer más legible el código. Como ejemplo tenemos el módulo `numpy` y sar el objeto `np` que se ve como sigue:

In [14]:
import numpy as np

### El famoso 'hola mundo' en python



In [15]:
print('Hola mundo')

Hola mundo


###  Tipos de variables

#### None type
`None`es comunmente utilizado como el resultado de una función que falló 

In [16]:
a = None
print(type(a))

<class 'NoneType'>


In [17]:
type(a)

NoneType

### Keyword `is`

Usamos `is` para verificar, por ejemplo, si dos variables son del mismo tipo

In [18]:
a = 5
b = 3.14
c = 10000.

type(a) is type(c)

False

In [19]:
type(a) == type(c)

False

In [20]:
type(b) is type(c)

True

### Listas, tuplas e índices

Al programar, vamos a querer usar objetos que guarden múltiples variables como arreglos (piensa en la series de tiempo financieras). Las **listas** y las **tuplas** te permiten hacer eso. Para identificarlas, las primeras usan `[]`y las segundas `()`

In [21]:
mi_lista = [3,1,4,1,6]
mi_tupla = (3,1,4,1,6)

Para acceder al los elementos de la lista o la tupla usamos su índice

In [22]:
print("El primer elemento de mi lista es", mi_lista[0])
print("El segundo elemento es", mi_lista[1])
print("El último elemento de mi tupla es", mi_tupla[-1])
print("El penúltimo elemento es", mi_tupla[-2])

El primer elemento de mi lista es 3
El segundo elemento es 1
El último elemento de mi tupla es 6
El penúltimo elemento es 1


La forma de indizar los elementos de una lista, tupla o cadena de elementos cualquiera (por ejemplo, la cadena $\mathbb{C}1\varepsilon(\eta)ci\alpha\int$ ) en python es iniciando con el cero y sumando uno por cada elemento al indizar de izquierda a derecha, o iniciando con el -1 y sumándole -1 a cada elemento si se indiza de derecha a izquiera

<img src="./img/indices_en_ciencias.png" style="width:400px">

Para acceder a más de un elemento de la lista o tupla hacemos lo siguiente

`[ primer_indice : ultimo_indice+1 ]` 

o también podemos obtener todos los elementos entre dos índices cada *n* pasos 

`[ primer_indice : ultimo_indice+1 : numero_de_pasos ]`

In [23]:
print("Los primeros dos elementos de mi lista son", mi_lista[0:2])
print("También, los primeros dos elementos de mi lista son", mi_lista[:2])

print("Los últimos dos elementos de mi lista son", mi_lista[-2:])
print("Los cinco primeros elementos de mi lista pero cada 2 pasos (uno sí y uno no) son",  mi_lista[0:5:2])
mi_lista

Los primeros dos elementos de mi lista son [3, 1]
También, los primeros dos elementos de mi lista son [3, 1]
Los últimos dos elementos de mi lista son [1, 6]
Los cinco primeros elementos de mi lista pero cada 2 pasos (uno sí y uno no) son [3, 4, 6]


[3, 1, 4, 1, 6]

Tanto las listas como las tuplas pueden guardar diferentes tipos de datos

In [24]:
mi_otra_lista = [100,True,3.1416]
print("El primer elemento es", type(mi_otra_lista[0]))
print("El segundo elemento es", type(mi_otra_lista[1]))
print("El tercer elemento es ", type(mi_otra_lista[2]))

El primer elemento es <class 'int'>
El segundo elemento es <class 'bool'>
El tercer elemento es  <class 'float'>


La **diferencia** entre una lista y una tupla es que las listas son mutables y las tupplas no. Una vez que creas una tupla, no puedes modificar ese objeto, por el contrario, una lista sí puede ser modificada incluso después de haber sido creada.

Podemos **añadir** elementos a `mi_lista`

In [25]:
mi_lista

[3, 1, 4, 1, 6]

In [26]:
mi_lista.remove(1) # Como argumento el elemento que queremos borrar
print(mi_lista)

[3, 4, 1, 6]


In [27]:
mi_lista.append(5) # Como argumento el elemento que queremos añadir
print(mi_lista)

[3, 4, 1, 6, 5]


In [28]:
mi_lista.insert(100,9) # Primer argumento el índice y el segundo argumento el elemento a añadir
print(mi_lista)

[3, 4, 1, 6, 5, 9]


<br><font color = #BB9963 size=5> ¡ojo! </font> Cuando hacemos una copia de una lista, lo que cambiamos en una, se cambia en las dos!!!

In [29]:
mi_lista = [8,9,7] 
mi_copia = mi_lista
mi_copia[0] = 10
print(mi_copia)
print(mi_lista)

[10, 9, 7]
[10, 9, 7]


A continuación cómo podemos remediar esto

In [30]:
mi_copia

[10, 9, 7]

In [31]:
import copy

mi_lista = [8,9,7]
mi_copia = copy.deepcopy(mi_lista)
mi_copia[2] = 6
print(mi_copia)
print(mi_lista)

[8, 9, 6]
[8, 9, 7]


### Ciclos

Los *ciclos* son operaciones fundamentales en programación. El ejemplo siguiente es un ciclo sobre un arreglo.

Presta atención en la identación, es la única manera de saber en dónde acaba el ciclo y continúa más código...

In [32]:
print(range(10))

range(0, 10)


In [33]:
for i in range(10):
    print("Número de ciclo", i)

    
print("-------")
print("Siempre iniciará en 0!!!")

Número de ciclo 0
Número de ciclo 1
Número de ciclo 2
Número de ciclo 3
Número de ciclo 4
Número de ciclo 5
Número de ciclo 6
Número de ciclo 7
Número de ciclo 8
Número de ciclo 9
-------
Siempre iniciará en 0!!!


Por otra parte, `i`, fuera del ciclo conservará el último valor que le fue asignado

In [34]:
enumerate(['¡que ondi!',1,True,1000])

Más ejemplos de ciclos, un ciclo usando `enumarete` y los ciclos escritos en una sola linea

In [35]:
for element in ['¡que ondi!',1,True,1000]: 
    print(element)
print("-------")

for (index,element) in enumerate(['¡que ondi!',1,True,1000]): 
    print("El elemento",index,"en la lista es:",element)
print("-------")

¡que ondi!
1
True
1000
-------
El elemento 0 en la lista es: ¡que ondi!
El elemento 1 en la lista es: 1
El elemento 2 en la lista es: True
El elemento 3 en la lista es: 1000
-------


In [36]:
mi_lista = [1,2,3,4,5]
# [1,4,9,16,25]

mi_lista_cuadrada = [element**2 for element in mi_lista]
print(mi_lista_cuadrada)

[1, 4, 9, 16, 25]


¿Cuál es el código largo para escribir lo mismo?

In [37]:
mi_lista = [1,2,3,4,5]
mi_lista_cuadrada2 =[]

for element in mi_lista:
    mi_lista_cuadrada2.append(element**2)
    
print(mi_lista_cuadrada)
print(mi_lista_cuadrada2)
    

[1, 4, 9, 16, 25]
[1, 4, 9, 16, 25]


### Expresiones condicionales

Típicamente, las condiciones son `<, >, ==, !=`

In [38]:
a = 10

if a > 5:
    print("la variable 'a' es mayor que 5")
else:
    print("la variable 'a' es menor o igual que 5")

b = 5
if b > 5:
    print("la variable 'b' es mayor que 5")
else:
    print("la variable 'b' es menor o igual que 5")

la variable 'a' es mayor que 5
la variable 'b' es menor o igual que 5


In [39]:
a = 10
x = a/10 if a > 5 else a
print(x)

b = 5
x = b/10 if b > 5 else b
print(x)

1.0
5


***

## Funciones en Python

Python te permite definir funciones, colocarle argumentos, ejecutar una tarea solicitada y devolver un resultado. La síntaxis de una función en python es

In [40]:
def mi_funcion_suma(x, y):
    """Esta función devuelve la suma de los dos argumentos
    x: primer argumento
    y: segundo argumento
    
    output: suma de x y y
    """
    return x+y

Es importante escribir qué hace la función y que argumentos recibe o la naturaleza de ellos para hacer más entendibe el código.

Una vez que se corrió la función, puede ser utilizada en cualquier momento.

In [41]:
mi_funcion_suma(10,2)

12

Otro ejemplo de función: 

La suma de los primeros n números naturales

In [42]:
def sum_integ(n):
    res = 0 # Inicializamos la variable res
    for i in range(1, n+1): 
        res+=i # Operador de suma y asigna
    return res

# Alternativamente podemos escribir la función anterior en un par de lineas
def sum_integ_alternative(n):
    return sum([i for i in range(1, n+1)])

# Y aquí con la fórmula teórica
def sum_integ_formula(n):
    return n*(n+1)/2

Recuerda, para cada entero $n$:
$$
\sum_{i=1}^{n} i  = \frac{n(n+1)}{2}
$$

In [43]:
n = 10

print("Suma de los primeros n números:", sum_integ(n))
print("Suma (alternativa) de los primeros n números:", sum_integ_alternative(n))
print("Fórmula teórica de las suma de los primeros n números:", sum_integ_formula(n))

Suma de los primeros n números: 55
Suma (alternativa) de los primeros n números: 55
Fórmula teórica de las suma de los primeros n números: 55.0


Una función puede tener varios argumentos o tomar argumentos que son keywords:

In [44]:
def my_function(x, a=1, b=0):
    return a*x+b

# Analicemos qué hace my_function en cada caso
print(my_function(1))
print(my_function(1, 2))
print(my_function(1, 1, 2))
print(my_function(1, b=2))

1
2
3
3


### Ejercicios

1. Dada una lista de enteros, por ejemplo a = [1,4,9,16,25,36], escribe una función de Python que tome como argumento esa lista 'a' y devuelva una lista que tenga únicamente sus elementos pares.

In [45]:
def get_even_numbers(input_list):
    """ Esta función devuelve los numeros pares de una lista 
    input: lista de números enteros
    output: lista de números pares
    
    """
    # Tu código va a continuación
    even_numbers_list = []
    for elemento in input_list:
        if elemento%2==0:
            even_numbers_list.append(elemento) 
    
    return even_numbers_list

# Forma alternativa de escribir la función
def get_even_numbers_bis(input_list):
    return [elemento for elemento in input_list if elemento%2==0]

In [46]:
lis=[1,4,9,16,25,36]

print(get_even_numbers_bis(lis))

[4, 16, 36]


Si corro estas dos líneas:

`lis=[1,4,9,16,25,36]`

`print(get_even_numbers(lis))`

Yo espero ver como output:

`[4,16,36]`

2. Crea una función que reciba un número entero y devuelva todos los divisores de ese número en una lista. 

In [47]:
def get_divisors(input_integer):
    """Esta función devuelve todos los divisores de un número dado
    input: integer
    output: list of divisors
    
    """
    return [elemento for elemento in range(1,int(input_integer/2)+1) if input_integer%elemento==0]

In [48]:
get_divisors(2022)

[1, 2, 3, 6, 337, 674, 1011]

Si corro esta línea:

`get_divisors(2022)`

Yo espero ver como output:

`[1, 2, 3, 6, 337, 674, 1011]`

3. Escribe una función que determine si un número es primo (número que no tiene divisores) o no. Hint: Usa la función get_divisors().

In [49]:
def is_prime(input_integer):
    """ Esta función devuelve True si el numero es primo, False si no es primo
    input: un entero
    output: un booleano (tipo de dato lógico)
    
    """
    
    return True if len(get_divisors(input_integer))==1 else False

In [50]:
is_prime(2022)

False

In [51]:
is_prime(13)

True

Si corro esta línea:

`is_prime(2022)`

Yo espero ver como output:

`False`

Pero si corro la línea

`is_prime(13)`

Yo espero ver como output:

`True`

4. Dadas dos listas, por ejemplo a = [5,2,87,23,45,8] y b = [4,3,5,7,8,42,65,8], escribe una función que devuelva una lista que contenga solo los elementos que se encuentran en las dos listas (elementos en común). Tu código debe funcionar para listas de distinto tamaño.

In [52]:
# Aquí va tu código

Para las listas del ejemplo `a = [5,2,87,23,45,8]` y `b = [4,3,5,7,8,42,65,8]`, espero ver como output:

`[8, 5]`

### Otros tips:

Si te lastima la retina el color blanco brillante de este tema de jupyter de notebook:

1. Puedes instalar **jupyterthemes** corriendo las siguientes lineas directamente en la línea de comdandos (o anaconda prompt) 

`python -m pip install jupyterthemes`

una vez que se instale todo, ejecutar el comando

`python -m pip install --upgrade jupyterthemes`

y posterior a esto con el comando `jt -l` apareceran las opciones de temas a elegir (por ejemplo: chesterish, grade3, gruvboxd, monokai, etc). Para elegir alguno ejecutas el comando 

`jt -t monokai`

Para regresar al tema por default de jupyter se utiliza el comando `jt -r`


2. De https://stackoverflow.com/questions/46510192/change-the-theme-in-jupyter-notebook, viene el siguiente consejo 

*Instead of installing a library inside Jupyter, I would recommend you use the 'Dark Reader' extension in Chrome (you can find 'Dark Reader' extension in other browsers, e.g. Firefox). You can play with it; filter the URL(s) you want to have dark theme, or even how define the Dark theme for yourself.*


3. Personaliza los colores que presentan tu patalla, en Windows es en el menú Configuración de Windows > Personalización > Filtros de color


### Relacionado con GitHub

Recomiendo que vean el siguiente video *Git and GitHub explained for beginners* en la liga youtube.com/watch?v=8Dd7KRpKeaE para saber más de la herramienta.